### ISSUE TITLE
Divisibility constraints are not propagated through symbolic expressions in Inductor

### ISSUE THREAD

**Issue by liqiangxl** [pytorch/pytorch#177146]
## Summary

`torch._check(x.shape[-1] % 16 == 0)` improves `torch.compile(dyanmic=True)` performance for RMS Norm, see #175755.
However, the divisibiliy is not propagated  to other symblic expres such as `xnumel = shape[0] * shape[-1]` in pointwise kernels. 
This causes significant performance loss when normalization kernel is segmented into `reduction + pointwise`.

## Root Cause

`statically_known_multiple_of(s27*s77, 16)` in `torch/_inductor/sizevars.py` asks sympy to evaluate `Eq(Mod(s27*s77, 16), 0)`. Even though the ShapeEnv has the axiom `Eq(Mod(s27, 16), 0)` (from the `torch._check`), sympy cannot deduce that a product `s27*s77` is divisible by 16 when one of its factors `s27` is — it does not perform the algebraic step: **if `a % n == 0` then `(a*b) % n == 0`**.

## Repro (on GB200)

```python
import torch

@torch.compile(dynamic=True, backend="inductor")
def f(x, w):
    torch._check(x.shape[-1] % 16 == 0)  # hidden_dim is always aligned
    rms = torch.sqrt(torch.mean(x.float() ** 2, dim=-1, keepdim=True) + 1e-6)
    return torch.nn.functional.gelu((x.float() / rms) * w.float()).half()

x = torch.randn([127, 262144], device="cuda", dtype=torch.float16)
w = torch.randn([262144], device="cuda", dtype=torch.float16)
f(x, w)
```

The graph segments into:
1. **Reduction kernels** (`2 kernels`): computes mean along `dim=-1` — `rnumel = shape[-1]` gets `tt.divisibility=16` correctly
2. **Pointwise kernel** (`1 kernel`): normalize + scale + GELU — `xnumel = shape[0] * shape[-1]` does **not** get `tt.divisibility=16`

The pointwise kernel's `xnumel` argument missing `tt.divisibility` prevents Triton from optimizing loads/stores, causing a **1.66x slowdown** on the pointwise kernel.

## Proposed Fix (Immediate)

In `statically_known_multiple_of`, when the direct sympy check fails and `numerator` is
a `sympy.Mul`, decompose it into factors and check if any individual factor is known to
be divisible by the denominator.

```diff
--- a/torch/_inductor/sizevars.py
+++ b/torch/_inductor/sizevars.py
@@ -417,7 +417,22 @@ class SizeVarAllocator:

         expr = sympy.Eq(Mod(numerator, denominator), 0)
-        return self.statically_known_true(expr)
+        if self.statically_known_true(expr):
+            return True
+
+        # If numerator is a product (a * b * ...), check if any factor is a
+        # multiple of the denominator. Handles e.g. xnumel = shape[0] * shape[-1]
+        # when torch._check(x.shape[-1] % 16 == 0) is asserted.
+        if isinstance(numerator, sympy.Mul) and isinstance(denominator, (int, sympy.Integer)):
+            denom_int = int(denominator)
+            for factor in numerator.args:
+                if isinstance(factor, (int, sympy.Integer)):
+                    if int(factor) % denom_int == 0:
+                        return True
+                else:
+                    factor_expr = sympy.Eq(Mod(factor, denominator), 0)
+                    if self.statically_known_true(factor_expr):
+                        return True
+
+        return False
```

## Benchmark Results (GB200, SM 10.0)

`[127, 262144]` fp16, RMSNorm + GELU with `torch._check(x.shape[-1] % 16 == 0)`:

| Kernel | Before fix | After fix | Speedup |
|--------|-----------|-----------|---------|
| reduction (triton_red) | 39.4 us | 34.3 us | 1.15x |
| persistent (triton_per) | 1.9 us | 1.4 us | 1.36x |
| **pointwise (triton_poi)** | **108.7 us** | **65.3 us** | **1.66x** |
| **Total** | **150.0 us** | **101.1 us** | **1.48x** |

### Versions

PyTorch version: 2.12.0a0+git5710dd2
Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (aarch64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jan 22 2026, 20:57:42) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-1008-nvidia-64k-aarch64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.2.46

cc @chauhang @penguinwu

**Comment by liqiangxl** [comment]
> Very interesting. Just wondering, does [your patch](https://github.com/pytorch/pytorch/pull/175755#issuecomment-3967378686) (which handles `divisible_by_16` automatically) need these rules too? Or, would we get the equivalent results by applying the patch?

1. We get the equivalent results by applying that patch. The issue with that patch is recompilation.

2. [Another fix ](https://github.com/pytorch/pytorch/pull/177061)is we generate two triton kernels for a segment, one fast path with div16 hints, the other general path without div16 hints. It can also solve the performance issue. Its problem is compile two kernels which increased cold-start time.
